Plenary template adjustment to fix the template library support

jouvin opened the pull request:
For host, cluster and metacluster plenary templates, include a preface template, called `archetype/declarations.pan`, intended mainly to define the loadpath for the rest of the profile. This happens only if the broker config option `object_declarations_template` is `true` (`false` by default).

Broker option `include_pan` is also removed (addition reverted), as it is superseded by the previous change

**This PR is critical for enable the use of the template library in Aquilon**

Fixes #86

jouvin commented:
@urbonegi I just updated the PR based on our discussion. I added another change I had in mind (I mentioned it during our discussion but it was may not clear) to allow to define that a variable is equal to another variable and I used this to define `LOADPATH` so that it is more clear that the initial value is the archetype.

Shaeli commented:
Hi, just noticed that all the commits in this PR have different change-id. Can you please change this to have the same change id? Thanks

jouvin commented:
@Shaeli Done! Sorry it was not clear for me if there should be one `Change-Id` for all the commits in the PR or one per commit...

jrha commented:
Please ignore the test failures, these are an artifact of the move from travisci.org to travisci.com

jouvin commented:
@jrha sorry I overlooked IDE-related templates... should be ok now.

jouvin commented:
@Shaeli I just fixed a typo in the second commit message... (missing `s`)

jrha commented:
I have just re-enabled branch protection for **upstream** and **master**, please let me know if it causes any problems pushing to the repo.

urbonegi commented on lib/aquilon/aqdb/model/building.py:
```diff
@@ -34,8 +34,4 @@ class Building(Location):
 
     address = Column(String(255), nullable=False)
 
-    next_rackid = Column(Integer, nullable=False, default=2)
```
Added by accident? Probably you need to rebase on latest upstream.

jouvin replied:
This is a little bit strange, I thought I just reverted the previous commit which had something related to building.py. I started from the latest upstream if I'm right (latest=1 week ago!). May be the easiest is that I remove this line (or building.py) from the revert commit if this is the only change?

jouvin commented:
@urbonegi @Shaeli Revert commit amended to remove `building.py` from it (was really part of the original commit reverted...).

urbonegi commented:
@jouvin so we do not understand fully why the archetype/base does not indeed become the first one if one would skip loading the pan/unis and functions? Please see my comment here:https://github.com/quattor/aquilon/issues/86

Another this is that, I advised @Shaeli that we do need same 'change-id' for all the commits in the PR - while it is not true. I got confused sorry. The change-id needs to be different for each commit. It is there so that gerrit can cope with rebase, when commit-id will change - change-id will not :) sorry for confusion!

jouvin commented:
@urbonegi I hope that https://github.com/quattor/aquilon/issues/86#issuecomment-392211167 answers your question! It seems that MS has some concerns (that I don't really understand frankly) about these changes, because they are not useful for MS who does not rely on the template library. To move forward with this PR asasp, as it is critical for allowing the use of Aquilon with the template library without hacking Aquilon, I propose to add a new config option, `template_library_configured`, that will be false by default and true in `conf.noms`. If this option is false, the generated profile will be unchanged, If it is true, the 2 changes implemented by this PR will be present in the profile. Would it help to have MS agreeing on this PR and get it through quickly?

About Gerrit Change-ID, don't worry. This is easy to generate again a separate one for each commit!

jouvin committed:
Revert "Add support for changing template LOADPATH"

- Except lib/aquilon/aqdb/model/building.py that should not have been part
  of the initial commit

This reverts commit 1ada34baa258498b6e2e81808ae9501cc44e3c5c.

Change-Id: I32e0681660f53db630cbcddf0163cb4e694d9ade

jouvin commented:
@urbonegi I implemented what I suggested in my previous comment, with even more elementary commits (support for variable definition based on a variable and addition of the new variable `ARCHETYPE` splitted in 2 commits).
New `host.py` successfully tested at LAL (we don't have `cluster` or `metacluster` objects but the change is trivial enough that I'm confident it is good!).

jouvin commented:
After the very long and intense discussion with @ned21 in #86 (😄), I remove from this PR the 2 commits not related to the template library support but to the addition of an `ARCHETYPE` variable. In the last version, the `archetype/loadpath` template **must** exist if the `template_library_configured` option is true, as per @ned21 suggestion.
2 minor details need to be agreed:
* Template name: @ned21 would prefer `declaration.pan` rather than `loadpath.pan` to make it clear that it should be a declaration template, as this is not possible to impose it in the object template. I tend to prefer a name that reflects the content expected and let to the documentation the task to explain that it should be a declaration template rather than a standard template (that would work anyway).
* Config option name: the idea was that this template is really required when you want to rely on templates that are imported from somewhere else and cannot be found in default include paths (plenary root, template-king repo...). This is typically the case with the template library (see https://github.com/quattor/release/tree/master/src/scripts/plenary_template_library), thus the name. The idea was to have an option that can be used for any object template variant that may be required by the template library use, if any other...

jrha commented:
For the record, I have no strong preferences for the name, as long as it is clear what it is for.

Ours currently looks like this:
```pan
declaration template archetype/preface;

final variable ARCHETYPE = "ral-tier1";

final variable QUATTOR_RELEASE = '17.12.0';

variable LOADPATH = prepend(SELF, format('template-library/%s/core', QUATTOR_RELEASE));
variable LOADPATH = prepend(SELF, format('template-library/%s/standard', QUATTOR_RELEASE));
variable LOADPATH = prepend(SELF, 'ral-tier1/core');
variable LOADPATH = prepend(SELF, 'ral-tier1/standard');
```

urbonegi commented on etc/aqd.conf.defaults:
```diff
@@ -345,8 +345,6 @@ gzip_output = false
 # Assume the webserver will decompress transparently as needed.
 # only used if gzip_output = true
 transparent_gzip = true
-template_extension = .tpl
```
why this is here?

jouvin replied:
A mistake of mine! Thanks for noticing it.

urbonegi commented:
Ok, this been on-going for some time now. We are happy to merge this one nearly as it is. We suggest to have code like this in hosts, clusters and metaclusters before the base/archetype temaple:

```
# Allow settings such as loadpath to be modified by the archetype before anything else happens
# Included only if template_library_configured option is true
# This template is optional: included only if it exists
if self.config.getboolean("panc", "include_declarations_template"):
    # Also contain other declarations such as pan/units and pan/functions
    pan_include(lines, "archetype/declarations")
    lines.append("")
else:
    pan_include(lines, ["pan/units", "pan/functions"])
    lines.append("")
# Okay, here's the real content
```

This new 'archetype/declarations' template should accommodate @jrha  suggestion above + could contain pan/units and pan/functions. MS later could also move to it too, but for now it could be controlled with an option.

urbonegi commented on etc/aqd.conf.defaults:
```diff
@@ -345,6 +345,8 @@ gzip_output = false
 # only used if gzip_output = true
 transparent_gzip = true
+# Generate the template library variant of the plenary object template if true
+template_library_configured = false
```
As per my suggestion change this option name as this is not obvious name for it.

urbonegi commented on lib/aquilon/worker/templates/cluster.py:
```diff
@@ -29,7 +29,7 @@
 from aquilon.worker.templates.panutils import (StructureTemplate, PanValue,
                                                pan_assign, pan_include,
-                                               pan_append)
+                                               pan_include_if_exists, pan_append)
```
pan_include_if_exists not used - no need to include

urbonegi replied:
pan_include_if_exists not used - no need to include

urbonegi commented on lib/aquilon/worker/templates/metacluster.py:
```diff
@@ -28,7 +28,7 @@
 from aquilon.worker.templates.panutils import (StructureTemplate, PanValue,
                                                pan_assign, pan_include,
-                                               pan_append)
+                                               pan_include_if_exists, pan_append)
```
pan_include_if_exists not used - no need to include

urbonegi commented on lib/aquilon/worker/templates/metacluster.py:
```diff
@@ -117,7 +117,17 @@ def get_key(self, exclusive=True):
+            lines.append("")
+        
         pan_include(lines, ["pan/units", "pan/functions"])
+        lines.append("")
```
why we have this extra append here?

urbonegi commented on lib/aquilon/worker/templates/cluster.py:
```diff
@@ -188,7 +188,17 @@ def get_key(self, exclusive=True):
+            lines.append("")
+
         pan_include(lines, ["pan/units", "pan/functions"])
+        lines.append("")
```
why we have this extra append here?

urbonegi commented on lib/aquilon/worker/templates/cluster.py:
```diff
@@ -188,7 +188,17 @@ def get_key(self, exclusive=True):
         return CompileKey.merge(keylist)
 
     def body(self, lines):
+        # Allow settings such as loadpath to be modified by the archetype before anything else happens
```
Suggest change to:
```
# Allow settings such as loadpath to be modified by the archetype before anything else happens
# Included only if template_library_configured option is true
# This template is optional: included only if it exists
if self.config.getboolean("panc", "include_declarations_template"):
    # Also contain other declarations such as pan/units and pan/functions
    pan_include(lines, "archetype/declarations")
    lines.append("")
else:
    pan_include(lines, ["pan/units", "pan/functions"])
    lines.append("")
# Okay, here's the real content
```

urbonegi commented on lib/aquilon/worker/templates/host.py:
```diff
@@ -374,6 +374,13 @@ def body(self, lines):
         services.sort()
         provides.sort()
 
+        # Allow settings such as loadpath to be modified by the archetype before anything else happens
```
Suggest change to:
```
# Allow settings such as loadpath to be modified by the archetype before anything else happens
# Included only if template_library_configured option is true
# This template is optional: included only if it exists
if self.config.getboolean("panc", "include_declarations_template"):
    # Also contain other declarations such as pan/units and pan/functions
    pan_include(lines, "archetype/declarations")
    lines.append("")
else:
    pan_include(lines, ["pan/units", "pan/functions"])
    lines.append("")
# Okay, here's the real content
```

urbonegi commented on lib/aquilon/worker/templates/metacluster.py:
```diff
@@ -117,7 +117,17 @@ def get_key(self, exclusive=True):
         return CompileKey.merge(keylist)
 
     def body(self, lines):
+        # Allow settings such as loadpath to be modified by the archetype before anything else happens
```
Suggest change to:
```
# Allow settings such as loadpath to be modified by the archetype before anything else happens
# Included only if template_library_configured option is true
# This template is optional: included only if it exists
if self.config.getboolean("panc", "include_declarations_template"):
    # Also contain other declarations such as pan/units and pan/functions
    pan_include(lines, "archetype/declarations")
    lines.append("")
else:
    pan_include(lines, ["pan/units", "pan/functions"])
    lines.append("")
# Okay, here's the real content
```

jouvin commented:
@urbonegi about your comment https://github.com/quattor/aquilon/pull/87#issuecomment-392815449, I disagree. I don't see any reason why we should move the include of `pan/units` and `pan/functions` to this template. The templates under `pan/` in the template library where called like this because they are the basic thing you need to use a pan template whatever you want to do with it, using the template library, `profile_base` or not... I don't see any reason to change this right now. If you see good reasons to do so, I'd like that you open an issue so that we can discuss the idea and agree on it. I made a mistake putting different things in this PR, this is fixed now but I think we should stick on fixing the critical issue in #86.

I'll fix the other mistakes.

urbonegi commented:
@ned21 suggested that if we add this new 'archetype/declarations' template to be included as the very first one, we better add the pan/units and pan/functions includes to it with other declarations rather than leave in broker code. Then later we can potentially get rid of Aquilon broker code for units and  functions include and just move to the declarations template in MS too. Including pan/units and pan/functions is the only use case for such new 'archetype/declarations' template in MS currently.

jouvin commented:
@urbonegi I think @ned21 mentioned that but also agreed later that it should be discussed separately. I don't see any reason to couple them and at this stage, I am not convinced this is a good move as it requires the user to do more when for me, the objective is to have the user doing only the things that the broker cannot do in a sensible manner (or without too much complexitiy).

jouvin commented:
I updated the PR taking into account most of @urbonegi remarks. 2 points still need to be agreed:
- The preface template name. As said by @jrha, I think it must reflects what the template is intended for. For me, `declaration` doesn't qualify as it is more about a type of template than the purpose of the template. I agree with @ned21 that `preface` is too vague. And I'm convinced we should insist that this template should not be used for anything else than adjusting the `LOADPATH` so that the templates until `base` can be found. Presently, I kept `loadpath` as the name but this  is easy to change when we come to an agreement.
- The `aqd.conf` option name. I went for `object_preface_template` to reflect the option purpose, with the idea (opposite to my initial one) that it should be used only for this purpose.

ned21 commented:
**re: The name** My view is that we cannot stop people making changes other than LOADPATH manipulations so we should not name it something that me become laughable false in a few years as people find other clever uses for it. Hence "declarations" is exactly what it says it is: you can use it for the same things that PAN defines as as declaration, anything else and your template will be invalid. And that's enforced in the compiler.

**The config option name**: `object_preface_template`: that's a good name, although we might rename preface to whatever we finally decide to call it.

**Removal of pan/units,functions**: I suggested this because those lines are now superfluous and can be more accurately managed per-archetype in the preface template. In particular for us it means that for those those archetypes where we do not use the definitions in those files, we can *remove the lines altogether* which we regard as a nicely improvement. It also makes the object template much *cleaner* by removing a hardcoded dependency on files from the TL that are not relevant to many of our archetypes.

jouvin commented:
@ned21 we are not in strong disagreement on the reasoning, despite we have not yet converged on the conclusion 😄 
- **template name**: I agree that we cannot stop people from using it for something else but I think we should do our best to avoid this template to become another `base`. Thus my proposal to have a name reflecting this. If my analysis is wrong that there is nothing else that is site/archetype dependent and cannot wait `base`, I agree that this could become a problem in the future. I missed the `s` in your proposal that makes it a little bit better than ` declaration`. But I also recognize that the exact usage of this template is influenced by your other proposal about `pan/...` templates.

- **Config option name**: I'm fine with updating my last proposal with the actual name choosen, despite I chose this name because I thought it was abstract enough to be usable with whatever name we decide. It is an easy change to do, I'm happy to update the PR if others prefer.

- **Removal of pan/... templates**: I understand that you don't want the object template to include things that are not useful for any object... But as soon as you use Pan, you probably need at least `pan/functions` which defines `push` function for example. These particular templates were designed to be the kind of thing you may need as long as you want to use Pan, even for something else than a host profile. Anyway, I accept that this is discussable. My main point here is that we should really make an effort to stick to what  you requested at some point: 1 PR = 1 goal. This is like the archetype discussion (#88): it was a mistake of mine to have started both things in the same PR. I don't see an urgency to decide on this (where there is one for the preface template as without it we cannot expect any TL site to adopt Aquilon) and my proposal was to handle it separately. I recognize that my `loadpath` name proposal is a little bit a conclusion for this point and I'm ready to accept `declarations` to ensure that we don't mix 2 different issues here and concentrate first on all the issues that are showstoppers for new sites (or may require to document hacks in the documentation which is very bad IMO).

urbonegi commented:
@ned21 if you agree, i think we could go one step at the time. For now just allow to add the extra initial template if newly defined config option is set to true. Then later we might move the functions and units there. However, calling that new template 'archetype/loadpath' is to explicit. It locks the usage of this template for one usecase - define the loadpath var. However, lets not limit ourselves like that. Please change the name to 'archetype/declarations' or 'archetype/preface' as suggested above. the config option then should be 'object_declarations_template' or 'object_preface_template'.  I vote for 'declarations'. @jouvin there are 2 small other issues in the PR -will add comments. Hopefully we will have an agreement with this - it been ongoing for too long ;)

urbonegi commented on etc/aqd.conf.defaults:
```diff
@@ -345,6 +345,9 @@ gzip_output = false
 # Assume the webserver will decompress transparently as needed.
 # only used if gzip_output = true
 transparent_gzip = true
+template_extension = .tpl
```
it would make sense to fix the 'Revert' commit to not delete this rather than re-add it in here.

jouvin replied:
You mean that this line disappeared because of the revert commit? This is stange... That means that the original commit did some bad things! I'll double check and fix the revert commit in this case...

urbonegi commented on lib/aquilon/worker/templates/cluster.py:
```diff
@@ -188,7 +188,15 @@ def get_key(self, exclusive=True):
+        # Included only if template_library_configured option is true
+        # This template is optional: included only if it exists
+        if self.config.getboolean("panc", "object_preface_template"):
+            pan_include(lines, "archetype/loadpath")
```
I think you missed: lines.append("") after pan_include

jouvin replied:
I thought you asked me to delete it in a previous comment.. I'll readd it!

urbonegi commented on lib/aquilon/worker/templates/host.py:
```diff
@@ -374,6 +374,12 @@ def body(self, lines):
+        # Included only if template_library_configured option is true
+        # This template is optional: included only if it exists
+        if self.config.getboolean("panc", "object_preface_template"):
+            pan_include(lines, "archetype/loadpath")
```
I think you missed: lines.append("") after pan_include

urbonegi commented on lib/aquilon/worker/templates/metacluster.py:
```diff
@@ -117,7 +117,15 @@ def get_key(self, exclusive=True):
+        # Included only if template_library_configured option is true
+        # This template is optional: included only if it exists
+        if self.config.getboolean("panc", "object_preface_template"):
+            pan_include(lines, "archetype/loadpath")
```
I think you missed: lines.append("") after pan_include

jouvin commented:
@urbonegi hopefully the last version of the commits should fix all your remaining remarks. Sorry for overlooking that the `aqd.conf.defaults` problem was introduced by the revert commit. I fixed it.

ned21 commented:
> if you agree, i think we could go one step at the time.

Sure. Where does that leave the code we added last year as an unsuccessful attempt to fix the same problem? If no one is using that, can we repurpose it to remove those options as a separately configurable item? Am I correct in thinking that at present the option replaces the include with an include { if_exists } ?

jouvin commented:
@ned21 yes the commit adding the `include_pan` option was removed (in fact the commit adding it was reverted, this is the first commit in this PR). It was a request by @urbonegi who said that this option was added only for the non MS users and thus had no longer any use. It is not a problem to readd it, in the view of the possiblity to have profiles without `pan/...` templates included in the object profile. If you confirm that the previous option was not used at MS, I'd suggest that we readd it after renaming it `include_pan_core_templates` as I find `include_pan` very misleading...

ned21 commented:
OK, so if it's reverted that's good. Re-adding it as include_pan_core_templates and have it not include them at all (not even with if_exists) is fine with me but I shall defer final judgement @urbonegi since she made a very good point about separation of concerns and tackling one thing in each PR.

jouvin commented:
So I think we all agree eventually! It was me who suggested first to separate both issues so I agree with @urbonegi remark on this point 😄 It'd be good if you (@ned21 ) could open another issue to tackle the pan core templates include. If @urbonegi agrees with the current state of this PR, you have the green light from my side to merge it after the internal review.

urbonegi commented on lib/aquilon/worker/templates/cluster.py:
```diff
@@ -188,12 +188,7 @@ def get_key(self, exclusive=True):
-            lines.append("")
-
-        # Okay, here's the real content
+        pan_include(lines, ["pan/units", "pan/functions"])
```
I am really confused to see this happened: hosts.py have the lines.append("") after the pan_include where cluster.py and metacluster.py do not...

jouvin replied:
@urbonegi the easiest is that you tell me what is your expectation for blank lines. My original PR was one after the include of `declarations.pan` and one after the include of pan core templates but you commented, if I understood properly, that it was too much. Is what is currently in `host.py` what you want everywhere?

urbonegi replied:
I see now that the host.py was inconsistent in blank lines initially (https://github.com/quattor/aquilon/commit/1ada34baa258498b6e2e81808ae9501cc44e3c5c) - so that is fine. Revert here without changing is fine.

urbonegi commented on etc/aqd.conf.lal:
```diff
@@ -0,0 +1,150 @@
+# This config file is for running the broker outside of MS where no afs available.
```
What is this :) ?

jouvin replied:
Another mistake of mine, should not be part of the commit....

jouvin committed:
Include an archetype preface template in the plenary object

- Only if object_declarations_template option is true: default=false
  (true in aqd.conf.noms)
- Template must be archetype/declarations (intended to define LOADPATH mainly)
- It is the first executed template in the plenary object
- Implemented for host, cluster and metacluster

Fixes #86

Change-Id: I7e57401fc82bfbc5af21d3d07efe543e999c23e6

jouvin commented:
@urbonegi I had another look to the commit, saw that the comment was still refering to `template_library_enabled` so I fixed it. I am not clear about your expecation regarding blank lines in the profile, as https://github.com/quattor/aquilon/pull/87#pullrequestreview-124047816 and https://github.com/quattor/aquilon/pull/87#pullrequestreview-124858080 seems a bit contradictory... I decided to put one blank line after the include of `declarations` and one after the include of pan core templates. Let me know if you prefer something different. At least it should be consistent now.

urbonegi commented:
@Shaeli it looks good now - ready to be fetched for review/internal CI. If tests succeed we can release this one next week.

Shaeli commented:
@urbonegi @jouvin Fetched for review and tests succeeded :)

jouvin commented:
Great!

urbonegi commented:
merged: https://github.com/quattor/aquilon/commit/3ccb23113d6efe174395a16cfbfd7175add820d5
